### DISCUSSION TITLE
Best way to fine tune LUT

### DISCUSSION THREAD

**Discussion by s1mptom** [awawa-dev/HyperHDR#916] [Q&A]
Hello,
I’m seeking advice on the best way (or application) to quickly and conveniently prepare a LUT file. I seem to be having some difficulties fine-tuning the picture using the built-in configurator. I haven’t been able to create a stable version where colors in both bright and dark scenes are displayed as close to accurate as possible. I’m using hyperion-webos as the image grabber on an LG G3 TV, with an Nvidia Shield as the source for Dolby Vision content. Adjusting the picture for bright scenes was relatively straightforward, but for some reason, dark scenes tend to shift towards blue-green tones. This might be artifacts from the grabber, or perhaps something else.
I would like to try preparing a LUT file with a bit more control over the settings, but other tools seem a bit too “professional” for my needs.

**Comment by awawa-dev** [comment]
Hi
This is no longer Dolby Vision but some signal converted in pain to RGB for internal webos needs. You would have to ask LG engineers why it came out so poorly or what webos capturing (powered by reverse engineering) is doing wrong. Maybe someday I will implement finally automatic calibration thanks to AI and the situation will improve **a little** but you should not expect miracles because we do not know what we are dealing with.

**Reply by s1mptom** [reply]
Yes, I understand that what I’m seeing doesn’t resemble unprocessed Dolby Vision (although SDR looks similar to what’s on the screen). It looks more like a simple RGB signal, but just very dark. Nevertheless, the picture isn’t that bad after applying the LUT. I’ve disabled the BT 2020 to 709 conversion, which partially helped get rid of the blue-green tint. That’s why I thought the picture could be improved—perhaps by removing some excess information in very dark areas, for example. I’d also like to experiment a little more…)
<img width="320" alt="image" src="https://github.com/user-attachments/assets/e72c3109-7371-4448-8c70-512ebe62108d"><img width="320" alt="image" src="https://github.com/user-attachments/assets/6528b49f-705a-41f6-b3f5-657518064f33">
<img width="320" alt="image" src="https://github.com/user-attachments/assets/c35e48fe-a974-4a2e-a4f2-a346df86c348"><img width="320" alt="image" src="https://github.com/user-attachments/assets/3429cc70-234e-4d18-909d-7f0d97d183d1">
<img width="320" alt="image" src="https://github.com/user-attachments/assets/1281ca99-8ef7-4fd8-b6b8-ca3cfc8c0dfe"><img width="320" alt="image" src="https://github.com/user-attachments/assets/6ca44943-9302-4850-8324-9d5fdee423e1">

**Reply by awawa-dev** [reply]
The colors look as if only gamma was missing (here you can guess which of the dozen or so most popular ones should be used). But gamma cannot disturb white balance, e.g. gray should not become bluish. If this happens on dark scenes, it may be due to rounding errors (then it is safest to cut it or make it a gray with averaged brightness) or it is not RGB but something very similar to it.

**Reply by s1mptom** [reply]
Maybe I would like to make the dark areas darker and reduce the color saturation in those areas (at least that’s what it seems like). As for the bright areas, I want to add more color saturation, though the bright areas aren’t as big of a problem. 
I think, it’s possible that image compression might be introducing some of its own artifacts or something like that )

Also, I completely forgot to thank you for your projects! Both HyperHDR and the firmware for the controller are fantastic. I really appreciate your work!

**Reply by awawa-dev** [reply]
> Maybe I would like to make the dark areas darker and reduce the color saturation in those areas (at least that’s what it seems like)

simplest solution that may work: if max(R,G,B) - min(R,G,B)  = 1 or <=2, then R=G=B=(R+G+B)/3, R/G/B are before any other correction. Should work especially well for darker scene.

> I think, it’s possible that image compression might be introducing some of its own artifacts or something like that 

they probably pack some higher scale (0-1024/10bit or even float) into 0-255 while including the target brightness component, so no all of the range is used. And because of that some important color components are lost or distorted.

> Also, I completely forgot to thank you for your projects! Both HyperHDR and the firmware for the controller are fantastic. I really appreciate your work!

Thank you! What tool do you use to modify LUT?

**Reply by awawa-dev** [reply]
> which partially helped get rid of the blue-green tint.

another possibility is: the source is in YUV but webos-capturing is using wrong coef to convert it to RGB

**Reply by s1mptom** [reply]
> What tool do you use to modify LUT?

I've only used the built-in LUT editing tool (from HyperHDR) for adjustments and haven't found a suitable alternative yet, which is why I posted my question here.

>they probably pack some higher scale (0-1024/10bit or even float) into 0-255 while including the target brightness component, so no all of the range is used. And because of that some important color components are lost or distorted.

Regarding the signal source, as far as I understand, hyperion-webos uses its own RAW(?) data grabber driver along with some converter to convert this RAW(?) data into a desired final format. I'm not very familiar with this, but I skimmed through the code. If it works as I think—taking some RAW(?) stream from the TV and converting it—there's a chance that the driver itself might be mishandling the format conversions. There's a line in the driver code that says:

```c
frame->pixel_format = PIXFMT_YUV420_SEMI_PLANAR; // ToDo: I guess?!
```

which raises some questions. And then there's the converter:

```c
else if (input->pixel_format == PIXFMT_YUV420_SEMI_PLANAR) {
    NV12ToARGB(
        input->planes[0].buffer,
        input->planes[0].stride,
        input->planes[1].buffer,
        input->planes[1].stride,
        output->planes[0].buffer,
        output->planes[0].stride,
        output->width,
        output->height);
}
```

This might explain some of the issues I'm experiencing.

**Reply by s1mptom** [reply]
In general, I dug around in the backend some more and tried intercepting a few frames as-is. It seems that the available LG API captures in NV12 with a loss of HDR metadata. At the moment, I don’t have any ideas on how to fix this. It looks like I’ll just have to accept it)

**Reply by awawa-dev** [reply]
Similar situation to USB capture cards and HDR10, but for grabbers we are exactly sure what the signal represents and we can detect proper YUV coefs. Here webos capturing takes some coef and converts YUV to RGB. HyperHDR receives such a RGB stream and can only correct gamma but will not correct coef because it is too late for that: it was already applied. You would have to send raw YUV signal to HyperHDR for calibration although of course it would require processing on both sides.

**Reply by s1mptom** [reply]
So, I can try to bypass the converter and send the stream directly to HyperHDR? Just in case, I’ll attach an archive with 2 frames that I captured from the built-in grabber.
[captured_frames.zip](https://github.com/user-attachments/files/16469888/captured_frames.zip)

**Reply by awawa-dev** [reply]
The Flatbuffers communication layer between webos capturing and HyperHDR currently only supports RGB format. That's why I wrote about the necessary changes.

**Reply by awawa-dev** [reply]
If we could add raw YUV sending (including NV12) there would be other advantages: HyperHDR LUT could perform conversion to RGB and tone mapping in one step, because one of the tables performs these two operations in one go. Then the conversion from RGB to YUV would be completely eliminated, which would probably relieve the system. How much depends on the resolution and capture frequency.

**Reply by s1mptom** [reply]
That certainly sounds great! However, I don’t really understand much about these topics, but I’m willing to dive in and try. As I understand it, we need to choose and implement a different frame transport method? I think if we eliminate the conversion in hyperion-webos itself, it would also positively impact speed and performance.

**Reply by awawa-dev** [reply]
This template should be extended for image format type on both sides:
https://github.com/awawa-dev/HyperHDR/blob/c6046581e0a3cdc9465582cb3bd12253f16cfc51/include/flatbuffers/parser/hyperhdr_request.fbs#L9
Then we can go to implement sending (I dont know webos capturing so cant help here) & decoding (NV12 is already supported but the flatbuffer server doesnt use it yet): https://github.com/awawa-dev/HyperHDR/blob/master/sources/utils/FrameDecoder.cpp
BTW small LUT table 50MB (instead of full one: 150MB) wont work anymore because it is missing the table we need for YUV & tone mapping decoding.

**Reply by s1mptom** [reply]
Oh, you’ve given me an interesting task/idea! I’ll dive deeper into how hyperion-webos and flatbuffers work and try to implement the sending of NV12. I’ll start by setting up a test server for debugging. As for the small LUTs—I don’t have any of those, only the large ones through the built-in calibration. I’ll be back when I have more questions :D

**Reply by s1mptom** [reply]
It looks like I've managed to send the original PIXFMT_YUV420_SEMI_PLANAR(?) image to my debug server. The image is, of course, dark even in SDR—probably because it's missing the conversion coefficients, but it's a start! To set up the stream reception on the HyperHDR side, do I just need to update the flatbuffers table, or is there anything else required?

My new table for sending (and receiving) looks like this:

```c
namespace hyperionnet;

// A priority value of -1 clears all priorities
table Register {
  origin:string (required);
  priority:int;
}

table RawImage {
  data:[ubyte];
  width:int = -1;
  height:int = -1;
}

table NV12Image {
  yData:[ubyte];
  uvData:[ubyte];
  width:int = -1;
  height:int = -1;
  stride_y:int = -1;
  stride_uv:int = -1;
}

union ImageType {RawImage, NV12Image}

table Image {
  data:ImageType (required);
  duration:int = -1;
}

table Clear {
  priority:int;
}

table Color {
  data:int = -1;
  duration:int = -1;
}

union Command {Color, Image, Clear, Register}

table Request {
  command:Command (required);
}

root_type Request;
```

Is using the NV12Image type correct? I noticed that the FrameDecoder uses just NV12—do I need to update this in the table and rebuild? Additionally, since I implemented this sending "on the side," parallel to the main one (going through the converter), I only implemented the frame sending and nothing else. Will this be sufficient, or do I need to implement any additional functions related to message processing on the HyperHDR side?

So, I want to understand what the next steps are that I need to take. Or if HyperHDR receives an NV12 image, will it handle everything else automatically?

**Reply by awawa-dev** [reply]
Great! Could you send me on email (the address is in my profile) size of the sample nv12 frame in bytes, width, height, stride_y, stride_uv? And if possible, an example of this nv12 frame in raw/binary form or encoded in flatbuffer format?

**Reply by s1mptom** [reply]
I’ve sent a bin file containing the y and uv data, hopefully to the correct address. Additionally, I’ve included the server log output with the message size and parameters (I hope it contains all the necessary information). It seems like the player “went to sleep” and reduced the brightness, so I’ll resend the captured frame just in case.
